src/target/riscv/riscv.c

```diff
@@ -1545,17 +1547,23 @@ int riscv_remove_watchpoint(struct target *target,
 /**
  * Look at the trigger hit bits to find out which trigger is the reason we're
  * halted.  Sets *unique_id to the unique ID of that trigger. If *unique_id is
- * ~0, no match was found.
+ * RISCV_TRIGGER_HIT_NOT_FOUND, no match was found.
  */
-static int riscv_hit_trigger_hit_bit(struct target *target, uint32_t *unique_id)
+
+static int riscv_hit_trigger_hit_bit(struct target *target, int64_t *unique_id)
 {
+	/* FIXME: this function assumes that we have only one trigger that can
+	 * have hit bit set. Debug spec allows hit bit to bit set if a trigger has
+	 * matched but did not fire. Such targets will receive erroneous results.
+	 */
```

Comment:
For what it's worth. hit bits are cleared by OpenOCD already. I've checked this visually. by inspecting the code.

Reply:
> I can see that the trigger hit bit is cleared on line 1615

Actually, I was talking about a different case. When we resume the target, we clear all hit bits anyway, since we temporally disable triggers when single-step is performed. If needed I can provide the exact lines. So technically we are fine for now.

> We believe the algorithm should continue clearing the "hit" bits of all triggers.

That being said, I agree that it would not hurt to continue the loop in this case.

Reply:
Thank you for explanation. I missed the fact that the temporary disabling of watchpoints on resume or step clears the "hit" bits as a by-product.

I would still prefer the clearing of all hit bits to occur in `riscv_trigger_detect_hit_bits()`, because:
- It is very unclear to the code reader that the clearing of "hit" occurs at a different place.
- We could make debug prints when more watchpoints have been hit - for easier troubleshooting.
- The disabling of triggers before step/resume is somewhat questionable because the debugger (GDB, LLDB) is attempting the same operation, so we are often duplicate in OpenOCD what the debugger is already attempting. For that reason, the disabling of triggers on step/resume may one day change - in which case we would lose the "hit" bit clearing (perhaps without even noticing).

In any case, this can be done in a separate PR.

Reply:
Just to wrap up this bit of  conversation (the part initiated by @JanMatCodasip)

I've looked at this in greater detail and here is my thoughts on the matter.

1. 

> I would still prefer the clearing of all hit bits to occur in riscv_trigger_detect_hit_bits(), because:

I don't think that `riscv_trigger_detect_hit_bits` should clear anything. It's job is to only detect hit bits. That's all. Clearing of hit bits should be handled elsewhere. Also I think we can make riscv_trigger_detect_hit_bits to detect all hit bits, not only the first one by allowing to call this function several times with a changed start index to make a lookup.

2.

> It is very unclear to the code reader that the clearing of "hit" occurs at a different place.

I agree that the way it is implemented now is confusing. I think that clearing of hit bits likely should happen in a dedicated function. 

3.

> The disabling of triggers before step/resume is somewhat questionable because the debugger 

It's not that questionable, actually. It is a matter of usability/convenience. In some debug scenarios it is not feasible to use high-level debugger like lldb or gdb because of side effects it has on a system (debugger loves to read memory here and there). Sometimes good-old telnet or tcl script is all we have.

Imagine a scenario when you hit a breakpoint or watchpoint. Now, you can read the processor state and want to move a little bit further using `step` command or even `resume`. If OpenOCD does not clear triggers - you'll end up stuck. In case of step you'll need to temporally disable the trigger. And you don't have a convenient interface to do that - you'll need to specify the virtual address than do the operation and then restore it! This is terrible for human and  inconvenient for scripting (since you'll need to parse text output of commands to get virtual address).  The  case of `resume` is similar (if not worse, since you won't have an opportunity to restore the trigger) - so the functionality to disable triggers has to stay.

That being said, RISC-V debug spec is incredibly quirky. For example it allows for various timings to be specified. Imagine that dreadful day when someone decides to implement **proper** support for timing "after"! In this case we may avoid disabling some triggers and we'll definitely require a dedicated functionality/function to clear the hit bit.

Reply:
(No need to block the merge request due to this open thread with broader discussion.)

Reply:
```
Would you know what other targets in OpenOCD do on TCL resume or step commands? Do they disable the breakpoints for the user or not?
```

AFAIK most of the targets to that. Last time I checked (at the time I've created https://github.com/riscv-collab/riscv-openocd/pull/741).  These were: arm, aarch64, mips, stm - all the major ones.

I don't see that other targets do any special processing with respect to trigger-like objects (maybe they don't need it? more on that below). Some targets have a code that handles masking of ISR, though. 

You can search it like this:

```
it grep handle_breakpoints | grep if | awk -F: '{ print $1 }' | sort | uniq
```

> What I was concerned (and considered questionable) is that riscv_openocd_step_impl() conditionally disables the current breakpoint (based on handle_breakpoints flag) but unconditionally disables all triggers - that doesn't seem consistent, and looks suspicious at the very least.

The reason is quite simple, I guess. For triggers you can't figure out which instruction is affected . I guess this is the reason.  Whether or not other target **require** such behavior for trigger-like entities is a separate question, since it heavily depends on how single-step is implemented by hw/architecture (like target architecture can explicitly state that watch points are not fired during hw-assisted single step if a certain bit is set).  It should be noted that the code that disabled triggers for RISC-V openocd was there for a while and I always took it for granted.

**To summarize**:

1. I think we can conclude that the removal of SW watchpoint at the current address is reasonable. And other targets do that.
2. As for other types of triggers... Now that you've mentioned this I kind of share the concern. I think we should at least investigate the behavior for other targets. And I think it would  not hurt to ask in OpenOCD mailing list regarding the expected semantics of 'step/resume' for the case of watchpoints. I'll create a separate OS issue.

Reply:
Thank you for the reply and I agree with your conclusions/summary.

Reply:
Ticket for further investigations: https://github.com/riscv-collab/riscv-openocd/issues/1079